src/daemon.rs

```diff
@@ -237,15 +267,18 @@ impl Future for DaemonRunner {
     type Error = Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        track!(self.http_server.poll())?;
+        let do_stop = track!(self.http_server.poll())?.is_ready() && self.do_stop;
+        if do_stop {
+            return Ok(Async::Ready(()));
+        }
         track!(self.rpc_server.poll())?;
         track!(self.rpc_service.poll())?;
-        let ready = track!(self.service.poll())?.is_ready();
-        if ready {
+        self.do_stop = self.do_stop || track!(self.service.poll())?.is_ready();
```

Comment:
NOTE: `self.service.poll` is ALWAYS called at least once by this tricky code even when `self.http_server.poll` finishes just after receiving a stop request.

Reply:
We get a lot of annoying logs by https://github.com/frugalos/frugalos/blob/c8468e63016d53068f0bc29716a63df237e45cd9/src/service.rs#L263 without `self.do_stop = self.do_stop || ...` condition.